src/functions.php

```diff
@@ -1,33 +1,118 @@
 <?php
 
-function getAllGames(object $db): array
+function getAllGames(PDO $db): array
 {
     $query = $db->prepare(
-        'SELECT `name`, `genre`.`genre`, `platform`.`platform`, `image`, `agerating`.`age` FROM videogames
+        'SELECT `name`, `genre`.`genre`, `platform`.`platform`, `image`, `agerating`.`agerating` FROM videogames
         JOIN `genre` ON `videogames`.`genreid` = `genre`.`id`
         JOIN `platform` ON `videogames`.`platformid` = `platform`.`id`
         JOIN `agerating` ON `videogames`.`ageid` = `agerating`.`id`;'
     );
     $result = $query->execute();
-    if ($result) {
-        $videogames = $query->fetchAll();
-    } else {
+    if (!$result) {
         throw new ErrorException("Couldn't retrieve information from database");
+    } else {
+        $videogames = $query->fetchAll();
     }
     return $videogames;
 }
 
+function checkImgExists(array $game): string {
+    if (!$game['image']) {
+        return "";
+    } else {
+        return "<img class='card-img' src='{$game['image']}' alt='Image of {$game['name']} cover'>";
+    }
+}
 function displayAllGames(array $games): string
 {
     $result = "";
     foreach ($games as $game) {
+        $img = checkImgExists($game);
         $result .= "<div class='card'>
-        <img class='card-img' src='{$game['image']}' alt='Image of {$game['name']} cover'>
+        {$img}
         <h2>Game: {$game['name']}</h2>
         <p><strong>Genre:</strong> {$game['genre']}</p>
         <p><strong>Platform:</strong> {$game['platform']}</p>
-        <p><strong>Age Rating:</strong> {$game['age']}</p>
+        <p><strong>Age Rating:</strong> {$game['agerating']}</p>
     </div>";
     }
     return $result;
+}
+
+function getTable(string $tablename, PDO $db): array {
+    $query = $db->prepare("SELECT `id`, `{$tablename}` FROM `{$tablename}`");
+    $result = $query->execute();
+    if (!$result) {
+        throw new ErrorException("Couldn't retrieve information from database");
+    } else {
+        return $query->fetchAll();
+    }
+}
+
+function setDropdownOptions(array $table, $column): string {
+    $options= '';
+    foreach($table as $row) {
+        $options .= "<option value=\"{$row['id']}\">{$row[$column]}</option>";
+    }
+    return $options;
+}
+
+function validateTitle(string $title) {
+    $title = trim($title);
+    if (empty($title) || strlen($title) > 100 || !preg_match("/^[a-zA-Z0-9\s:!]+$/", $title)) {
+        return false;
+    }
```

Comment:
Is a title with an underscore not valid? Hyphenated titles? Titles with Question marks? Titles with any other special character, like the infinity symbol? Titles in Japanese?

I think this RegEx is too restrictive for validating a videogame title. When it comes to strings, as long as you strip the harmful HTML stuff with `htmlspecialchars()` it should be considered 'allowed' unless there *actually is* some sort of accepted standard format - with names and titles, there are basically no rules.